csharp/src/Apache.Arrow/C/CArrowArrayStreamImporter.cs

```diff
@@ -0,0 +1,135 @@
+        /// IArrowArrayStream importedStream = CArrowArrayStreamImporter.ImportStream(importedPtr);
+        /// </code>
+        /// </examples>
+        public static unsafe IArrowArrayStream ImportArrayStream(CArrowArrayStream* ptr)
+        {
+            return new ImportedArrowArrayStream(ptr);
+        }
+
+        private sealed unsafe class ImportedArrowArrayStream : IArrowArrayStream
+        {
+            private readonly CArrowArrayStream* _cArrayStream;
+            private readonly Schema _schema;
+            private bool _disposed;
+
+            public ImportedArrowArrayStream(CArrowArrayStream* cArrayStream)
+            {
+                if (cArrayStream == null)
+                {
+                    throw new ArgumentNullException(nameof(cArrayStream));
+                }
+                _cArrayStream = cArrayStream;
+                if (_cArrayStream->release == null)
+                {
+                    throw new ArgumentException("Tried to import an array stream that has already been released.", nameof(cArrayStream));
+                }
+
+                CArrowSchema* cSchema = CArrowSchema.Create();
+                try
+                {
+                    if (_cArrayStream->get_schema(_cArrayStream, cSchema) != 0)
+                    {
+                        throw new Exception("This needs to be better");
+                    }
+                    _schema = CArrowSchemaImporter.ImportSchema(cSchema);
+                }
+                finally
+                {
+                    if (_schema == null)
+                    {
+                        CArrowSchema.Free(cSchema);
+                    }
+                }
+            }
+
+            ~ImportedArrowArrayStream()
+            {
+                Dispose();
+            }
+
+            public Schema Schema => _schema;
+
+            public ValueTask<RecordBatch> ReadNextRecordBatchAsync(CancellationToken cancellationToken = default)
+            {
+                if (_disposed)
+                {
+                    throw new ObjectDisposedException(typeof(ImportedArrowArrayStream).Name);
+                }
+
+                RecordBatch result = null;
+                CArrowArray* cArray = CArrowArray.Create();
+                try
+                {
+                    if (_cArrayStream->get_next(_cArrayStream, cArray) != 0)
+                    {
+                        throw new Exception("This too needs to be better");
+                    }
+                    if (cArray->release != null)
+                    {
+                        result = CArrowArrayImporter.ImportRecordBatch(cArray, _schema);
+                    }
+                }
+                finally
+                {
+                    if (result == null)
+                    {
+                        CArrowArray.Free(cArray);
+                    }
+                }
+
+                return new ValueTask<RecordBatch>(result);
```

Comment:
You could, potentially, accept a task scheduler and schedule a new task to call `get_next`, allowing this to be more accurately async.  Though that should definitely be a follow-up and probably depends on what you're interfacing with (e.g. is the underlying stream performing I/O and slow?)